woocommerce/woocommerce-ios
# Shipping Labels: Complete validation for customs form

Closes #4687 

# Description
This PR aims to complete validation of customs form with two parts:
### Customs form list Done button. 
To prepare for multi-package support of Shipping Labels, the customs screen was structured like so: Customs Form List -> Customs Form -> Item Details. 

For validation, we want to enable / disable Done button on the Customs Form List immediately when any field is updated. So the validation states have to be sent upward from a child item to its parent form, all the way to the containing form list.

### ITN field validation
It is required that that if total value of items that share the same HS tariff number exceeds $2,500, ITN field in the customs form is required.

To do this, each item has to let its parent know of the latest validation states of the item's value and tariff number. The parent will collect these states to calculate whether any tariff number valued more than $2,500, to decide whether ITN field should be required.

# Changes
### Item details view model
- Added new `@Published` variables for validated tariff number and total value of the item. These will be observed by the parent form to check which tariff number's value exceed $2,500 and requires ITN.
- Added ~~property observers~~ mappers for `value` and `hsTariffNumber` to update the validated total value and tariff number.
- Added new `@Published` variable `validItem` to let parent form know if all the fields in the item details are valid or not.
- Added private method helpers for validation, which are reused between calculated variables and overall validation check.

### Customs form input view model
- Added new `@Published` variable for `validForm` to let parent list know if all the fields as well as items in the form are valid.
- Observe all item view models' `validItem` to save the validation states for all items. These states are used to validate the form.
- Observe all item view models' `validatedHSTariffNumber` and `validatedTotalValue` and save them to configure a list of tariff numbers that require ITN.
- Added private method helpers for validation, which are reused between calculated variables and overall validation check.

### Customs form list view model
- Observe validation states of all child forms and save them locally.
- Configured Done button to enable only when all child forms are valid.
- Configured validated form list by fetching validated form from all input view models.

### Customs form list UI
- Return the validated form list when Done button is tapped.
- Disable Done button when it should not be enabled

### Update unit tests for view models.

# Demo
![customs-validation](https://user-images.githubusercontent.com/5533851/130007313-1641eb4b-74d3-4d46-9a39-f1abb798ab1d.gif)

# Testing steps
1. Make sure that your test store is eligible for creating shipping labels (has WCShip plugin installed and packages configured in WP Admin > WooCommerce > Settings > Shipping > Woocommerce Shipping).
2. Create an order that requires international shipping (origin country different from destination country).
3. Open app, navigate to Orders tab and select the created order.
4. Select Create Shipping Label.
5. Skip through Ship From, Ship To, Package Details to configure Customs.
6. Edit an item in the customs form to have a 6-digit tariff number, and a value of more than $2500 (or if there are more than 1 item, update all item to have same tariff number and total value more than $2500). Notice that there's an validation error in ITN field showing up asking for input.
7. Edit fields in the form to invalid values, notice that Done button is disabled. Update the fields to be valid, Done button should be enabled.

Update release notes:
